ibm/resource_ibm_securitygroup_rule.go

```diff
@@ -0,0 +1,250 @@
+		return nil, fmt.Errorf("Error fetching information for Security Group Rule: %s", err)
+	}
+
+	if len(rules) == 0 {
+		return nil, nil
+	}
+	return &rules[0], nil
+}
+
+func resourceIBMSecurityGroupRuleCreate(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ClientSession).SoftLayerSession()
+	service := services.GetNetworkSecurityGroupService(sess)
+
+	securitygroupid := d.Get("security_group_id").(int)
+
+	sgrule := datatypes.Network_SecurityGroup_Rule{}
+
+	if d.Get("direction").(string) != "" {
+		sgrule.Direction = sl.String(d.Get("direction").(string))
+	}
+
+	if d.Get("ether_type").(string) != "" {
+		sgrule.Ethertype = sl.String(d.Get("ether_type").(string))
+	}
+
+	if d.Get("port_range_min").(int) > 0 {
+		sgrule.PortRangeMin = sl.Int(d.Get("port_range_min").(int))
+
+	}
+
+	if d.Get("port_range_max").(int) > 0 {
+		sgrule.PortRangeMax = sl.Int(d.Get("port_range_max").(int))
+
+	}
+
+	if d.Get("protocol").(string) != "" {
+		sgrule.Protocol = sl.String(d.Get("protocol").(string))
+	}
+
+	// if only one of min/max is provided, set the other one to the provided
+	if sgrule.PortRangeMin != nil && sgrule.PortRangeMax == nil {
+		sgrule.PortRangeMax = sgrule.PortRangeMin
+	}
+	if sgrule.PortRangeMax != nil && sgrule.PortRangeMin == nil {
+		sgrule.PortRangeMin = sgrule.PortRangeMax
+	}
+
+	if sgrule.PortRangeMin != nil && *sgrule.PortRangeMax < *sgrule.PortRangeMin {
+		return fmt.Errorf("port range max should be greater or equal to min")
+	}
+
+	matchingrule, err := findMatchingRule(securitygroupid, &sgrule, service)
+	if err != nil {
+		return err
+	}
+
+	if matchingrule != nil {
+		log.Printf("[INFO] rule exists")
+		d.SetId(fmt.Sprintf("%d", *matchingrule.Id))
+		return nil
+	}
+
+	opts := []datatypes.Network_SecurityGroup_Rule{
+		sgrule,
+	}
+	log.Println("[INFO] creating security group rule")
+	_, err = service.Id(securitygroupid).AddRules(opts)
+	if err != nil {
+		return fmt.Errorf("Error creating Security Group Rule: %s", err)
+	}
+
+	matchingrule, err = findMatchingRule(securitygroupid, &sgrule, service)
+	if err != nil {
+		return err
+	}
+	d.SetId(strconv.Itoa(*matchingrule.Id))
+
+	return resourceIBMSecurityGroupRuleRead(d, meta)
+}
+
+func resourceIBMSecurityGroupRuleRead(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ClientSession).SoftLayerSession()
+	service := services.GetNetworkSecurityGroupService(sess)
+
+	matchingrules, err := service.Filter(filter.Build(
+		filter.Path("rules.id").Eq(d.Id()))).Id(
+		d.Get("security_group_id").(int)).GetRules()
+	if err != nil {
+		// If the group is somehow already destroyed, mark as
+		// succesfully gone
+		if err, ok := err.(sl.Error); ok && err.StatusCode == 404 {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error retrieving Security Group Rule: %s", err)
+	}
+
+	if len(matchingrules) == 0 {
+		d.SetId("")
+		return nil
+	}
+
+	if matchingrules[0].Direction != nil {
+		d.Set("direction", matchingrules[0].Direction)
```

Comment:
There is no need to check for the `nilness` as d.Set takes care of that. You could remove that. Same goes for the below code.
As you have already noted in that case there is no need to deference the pointer also.